### ISSUE TITLE
Using PyMODA functions in standalone scripts

### ISSUE THREAD

**Issue by spmccormack** [luphysics/PyMODA#10]
As suggested by @maximelucas, it would be useful to be able to import and use PyMODA's functionality - for example, the wavelet transform function - in standalone scripts.

**Comment by spmccormack** [comment]
Unfortunately, using MATLAB-packaged code is a slightly complicated process. 

[Linux support is problematic](https://github.com/luphysics/PyMODA/blob/master/docs/developer-guide.md#linux-support). This has caused PyMODA's wrapper functions to be tied in with other parts of the PyMODA codebase, making them more difficult to use in standalone programs.

Additionally, on Linux it is probably not possible to use Matplotlib in the same script that calls MATLAB code, because Matplotlib uses PyQt under the hood. There are 2 workarounds for this:

1. Perform calculations and save the data using one script, and plot the data using another script.
2. Perform all calculations in a separate process and only set the `LD_LIBRARY_PATH` [in that process](https://stackoverflow.com/a/56886473/6364450). This technique is used in PyMODA, and it can be used to speed up the execution of the program if you perform multiple separate calculations (e.g wavelet transforms) each on a separate process.

**Comment by spmccormack** [comment]
It should be possible to extract the PyMODA algorithms as a library which PyMODA and other programs would be able to import and call. 

This would make PyMODA simpler and allow easier usage of its algorithms in standalone scripts. However, one of the above workarounds would still be required on Linux.

@maximelucas Do you think this approach would be useful?

**Comment by pavleb** [comment]
I don't know if I undestand correctly, but you can use PyModa within Python quite simple. Put the maths folder in you python path and then do the following:

```
import maths.algorithms.wavelet_transform as wt

wp = wt.MorseWavelet(f0=1,a=3)
W,freq_w = wt.wft(sig,fs,wp,fmin=fb/10,fmax=fb,padmode=None,cut_edges=True)

fig, ax = plt.subplots(1, 1)
fig.set_size_inches(6, 2)
ax.pcolormesh(t,freq_w,np.abs(W),shading = 'goaround',rasterized=True)
```

I hope that was the think you were looking for.

**Comment by maximelucas** [comment]
> I don't know if I undestand correctly, but you can use PyModa within Python quite simple. Put the maths folder in you python path and then do the following:
> 
> ```
> import maths.algorithms.wavelet_transform as wt
> 
> wp = wt.MorseWavelet(f0=1,a=3)
> W,freq_w = wt.wft(sig,fs,wp,fmin=fb/10,fmax=fb,padmode=None,cut_edges=True)
> 
> fig, ax = plt.subplots(1, 1)
> fig.set_size_inches(6, 2)
> ax.pcolormesh(t,freq_w,np.abs(W),shading = 'goaround',rasterized=True)
> ```
> 
> I hope that was the think you were looking for.

This solution seems to work great. 
One comment for potential future readers: If I'm not mistaken, the function is `wt.wt()` and not `wt.wft()`.
One question though: as far as I understand, the file used in this solution, `/home/maxime/Documents/pyMODA/src/maths/algorithms/wavelet_transform.py` is a python translation of the original MATLAB script, rather than a wrapper, which is why it works in the same script as matplotlib. However, at the beginning of there is this line:
`'An attempt to convert the Matlab WT algorithm into Python. Not in use.'` So is this function safe to use, or is not finished yet?

This solution seems to work only for the functions that have been translated to python. For the functions that have not, yet, I guess @spmccormack's solution would be helpful. Although, I personally currently only need the wavelet transform.

**Comment by pavleb** [comment]
@maximelucas As you noticed, the name of the function is wrong. It is going to be fixed ASAP. The note of the function is also old. The majority of the numerical testing is completed. Currently, I am using this function quite actively in my research, so any numerical bugs are immediately fixed. I advise you to check for any new commits.

However, it should be stressed that Sam is the lead, I am just trying to support him in this limited area.

**Comment by spmccormack** [comment]
Thanks @pavleb!

@maximelucas If you do want to use the more tried-and-tested wavelet transform from MODA, here's a quick example:

---

`utils.py`

```python
import numpy as np
from numpy import ndarray
import os


def matlab_to_numpy(arr) -> ndarray:
    """
    Converts a matlab array to a numpy array.
    """
    try:
        # Faster for real arrays, raises exception for complex arrays.
        result = np.array(arr._data).reshape(arr.size, order="F")
    except:
        result = np.array(arr)
    return result

def enable_runtime() -> None:
    """
    [Only required for Linux.]

    Sets the LD_LIBRARY_PATH to make the MATLAB Runtime accessible.
    """
    # The long string "/usr/local/..." may be different; use the value shown by the MATLAB Runtime installer.
    os.environ[
        "LD_LIBRARY_PATH"
    ] = "/usr/local/MATLAB/MATLAB_Runtime/v96/runtime/glnxa64:/usr/local/MATLAB/MATLAB_Runtime/v96/bin/glnxa64:/usr/local/MATLAB/MATLAB_Runtime/v96/sys/os/glnxa64:/usr/local/MATLAB/MATLAB_Runtime/v96/extern/bin/glnxa64"
```

---

`example.py`

```python
import utils

utils.enable_runtime()  # This must be before "import WT".
import WT  # Import the "WT" MATLAB-packaged library.
import matlab  # This must be after "import WT".

import numpy as np
import scipy.io

# This is 10.0, rather than 10, to prevent issues with MATLAB's data casting.
fs = 10.0

# Create artificial signal.
t = np.arange(0, 100, 1/fs)
signal = np.sin(2*t + np.cos(t))

signal_matlab = matlab.double(
    signal.tolist()  # MATLAB can only convert lists, not Numpy arrays, so `tolist()` is required.
)

package = WT.initialize()
wt, freq = package.wt(
    signal_matlab, fs, "fmax", 1.0, nargout=2,  # `nargout` defines the number of returned variables.
)

wt_numpy = utils.matlab_to_numpy(wt)
freq_numpy = utils.matlab_to_numpy(freq)

from matplotlib import pyplot as plt

mesh1, mesh2 = np.meshgrid(t, freq_numpy)
plt.contourf(mesh1, mesh2, wt_numpy, 256)
plt.show()
```

---

The actual function call, `package.wt(...)` is essentially the same as in MATLAB; you just need to specify `nargout`, or MATLAB won't know how many arguments to return.

You have to do some awkward data conversions, and passing an int (e.g `2`) instead of a float (e.g `2.0`) often causes strange errors as MATLAB converts to an int instead of a double. For more information regarding how MATLAB converts data types, see [Pass Data to MATLAB from Python](https://www.mathworks.com/help/matlab/matlab_external/pass-data-to-matlab-from-python.html).

**Comment by maximelucas** [comment]
Great, thanks a lot to both.

@pavleb is `wt.wt()` from your solution computing the power or the amplitude for the wavelet transform? I couldn't find a function argument or any hint in the code.

Same question for your solution @spmccormack?

**Comment by spmccormack** [comment]
> Great, thanks a lot to both.
> 
> @pavleb is `wt.wt()` from your solution computing the power or the amplitude for the wavelet transform? I couldn't find a function argument or any hint in the code.
> 
> Same question for your solution @spmccormack?

---

`wt, freq = package.wt(..., nargout=2)` from the `WT` package will return the same values as the MATLAB function: 

```matlab
% OUTPUT:
% WT   - wavelet transform of the signal [sig],
%        with rows corresponding to frequencies and columns - to time;
%        represents FNxL matrix, where FN is the number of frequencies
%        and L is the length of the signal in samples
% freq - frequencies corresponding to rows of WT
```
This will be the amplitude, rather than the power.

---

`wt.wt(...)` should also return amplitude.

---

**Comment by spmccormack** [comment]
Hi all,

I've been working on this issue for a while, in the form of [PyMODAlib](https://github.com/luphysics/PyMODAlib). PyMODA is now using PyMODAlib for its wavelet transform, and I will be migrating the remaining algorithms across when I have time. 

PyMODAlib has Sphinx documentation (for example, here's the documentation for the [wavelet transform](https://pymodalib.readthedocs.io/en/latest/source/pymodalib.algorithms.wavelet.html#module-pymodalib.algorithms.wavelet)), and the README for the repository has a User Guide and a Developer Guide for getting started.

I've also created a new [DOI for PyMODAlib](https://zenodo.org/record/3730978).

---

@pavleb Thanks again for your improvements to the wavelet transform! I merged your commits from the `pavle_updates` branch before moving the wavelet transform to PyMODAlib, but if you have any more commits that haven't been pushed then let me know and I'll see that they reach PyMODAlib safely. 

By the way, I've invited you as a collaborator for the PyMODAlib repository.
